multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -81,129 +164,284 @@ static void _configureChannel(int dma, int channel, int dir, int priority, void
+static int libdma_hasChannelFinished(const void *maddr, volatile unsigned int *channel_base)
 {
-	return (maddr != NULL && *(channel_base + cndtr) > 0);
+	return (maddr != NULL) && (*(channel_base + cndtr) == 0);
 }
 
 
-static int _transfer(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
+static int libdma_transferHelper(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
 {
 	/* Empirically chosen value to avoid mutex+cond overhead for short transactions */
 	int use_interrupts = len > 24;
+	int interrupts_flags = (use_interrupts == 0) ? 0 : DMA_TCIE_FLAG;
+	volatile int rx_done = 0;
 
-	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + 5 * rx_channel;
-	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + 5 * tx_channel;
+	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + (5 * rx_channel);
+	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + (5 * tx_channel);
+
+	/* Only one request may be issued on one channel at one time. */
+	if ((dma_transfers[dma][rx_channel].type != dma_transferNull) ||
+		(dma_transfers[dma][tx_channel].type != dma_transferNull)) {
+		return -EINVAL;
+	}
 
-	if (rx_maddr != NULL)
-		_prepare_transfer(rx_channel_base, rx_maddr, len, use_interrupts);
+	dma_transfers[dma][rx_channel].type = dma_transferOnce;
+	dma_transfers[dma][rx_channel].once.done_flag = &rx_done;
+	dma_transfers[dma][rx_channel].once.channel_base = rx_channel_base;
+	libdma_prepareTransfer(rx_channel_base, rx_maddr, len, interrupts_flags);
 
-	if (tx_maddr != NULL)
-		/* When doing rw transfer, avoid unnecessary interrupt handling and condSignal()
-		by waiting only for RX transfer completion, ignoring TX */
-		_prepare_transfer(tx_channel_base, (void *)tx_maddr, len, use_interrupts && rx_maddr == NULL);
+	dma_transfers[dma][tx_channel].type = dma_transferOnce;
+	/* When doing rw transfer, avoid unnecessary interrupt handling and condSignal()
+	by waiting only for RX transfer completion, ignoring TX */
+	libdma_prepareTransfer(tx_channel_base, (void *)tx_maddr, len, 0);
 
-	if (use_interrupts) {
+	if (use_interrupts != 0) {
 		mutexLock(dma_common[dma].irqLock);
-		while (_has_channel_finished(rx_maddr, rx_channel_base) || _has_channel_finished(tx_maddr, tx_channel_base))
+		while ((rx_done == 0) || (libdma_hasChannelFinished(tx_maddr, tx_channel_base) == 0)) {
 			condWait(dma_common[dma].cond, dma_common[dma].irqLock, 1);
+		}
 		mutexUnlock(dma_common[dma].irqLock);
 	}
 	else {
-		while (_has_channel_finished(rx_maddr, rx_channel_base) || _has_channel_finished(tx_maddr, tx_channel_base))
+		while ((libdma_hasChannelFinished(rx_maddr, rx_channel_base) == 0) || (libdma_hasChannelFinished(tx_maddr, tx_channel_base) == 0))
 			;
+
+		dma_transfers[dma][rx_channel].type = dma_transferNull;
+		dma_transfers[dma][tx_channel].type = dma_transferNull;
```

Comment:
Already unconditionally assigned just below this block.

Reply:
fixed